db/migrate/20220815210513_add_hidden_check_to_repo_permissions_function.rb

```diff
@@ -0,0 +1,131 @@
+class AddHiddenCheckToRepoPermissionsFunction < ActiveRecord::Migration[7.0]
+  def up
+    execute %(
+DROP FUNCTION IF EXISTS check_repo_permissions(varchar, varchar, varchar);
+CREATE FUNCTION check_repo_permissions(user_name_ varchar, course_name varchar, repo_name_ varchar)
+RETURNS boolean
+LANGUAGE plpgsql
+AS
+$$
+DECLARE
+    role_type varchar;
+    role_id_ integer;
+BEGIN
+    SELECT roles.id, roles.type
+    INTO role_id_, role_type
+    FROM users
+        JOIN roles ON roles.user_id=users.id
+        JOIN courses ON roles.course_id=courses.id
+    WHERE courses.name=course_name AND users.user_name=user_name_ AND roles.hidden=false
+        FETCH FIRST ROW ONLY;
+
+    IF role_type IN ('Instructor', 'AdminRole') THEN
+        RETURN true;
+    END IF;
+    IF role_type = 'Ta' THEN
+        RETURN EXISTS(
+                SELECT 1
+                FROM memberships
+                    JOIN roles ON roles.id = memberships.role_id
+                    JOIN groupings ON memberships.grouping_id = groupings.id
+                    JOIN groups ON groupings.group_id = groups.id
+                    JOIN assignment_properties ON assignment_properties.assessment_id = groupings.assessment_id
+                WHERE memberships.type = 'TaMembership'
+                  AND assignment_properties.anonymize_groups = false
+                  AND roles.id = role_id_
+                  AND groups.repo_name = repo_name_
+            );
+    END IF;
+    IF role_type = 'Student' THEN
+        RETURN EXISTS(
+                SELECT roles.id
+                FROM memberships
+                    JOIN roles ON roles.id=memberships.role_id
+                    JOIN groupings ON memberships.grouping_id=groupings.id
+                    JOIN groups ON groupings.group_id=groups.id
+                    JOIN assignment_properties ON assignment_properties.assessment_id=groupings.assessment_id
+                    JOIN assessments ON groupings.assessment_id=assessments.id
+                    LEFT OUTER JOIN assessment_section_properties ON assessment_section_properties.assessment_id=assessments.id
+                WHERE memberships.type='StudentMembership'
```

Comment:
We should not allow students who are in a hidden course to access repositories. The best way to do this here is to join assessments table to the courses table and then add another where condition that checks if the course is hidden